src/ir/module.h

```diff
@@ -34,11 +34,13 @@ class Block;
 class Operation {
  public:
   Operation(const Block* parent, const Operator& op,
-            NamedAttributeMap attributes, NamedValueMap inputs)
+            NamedAttributeMap attributes, NamedValueMap inputs,
+            std::unique_ptr<Module> module = nullptr)
```

Comment:
One more thought about this `module`: if this corresponds to the implementation, shouldn't it be attached to the `Operator`?

Reply:
This allows more flexibility. For example, we don't have define and register an operator for every function or particle that we want to define. We will have an operator called `MakeFunction` (or `MakeParticle`) to define a function. The `Operation` (not `Operator`) instance of a `MakeFunction` will define the body of the function using the module. 

We will invoke the function with `CallFunction` (or `CallParticle`) that will take the `MakeFunction` operation instance as an argument (or perhaps attribute).

Reply:
That sounds like a level of dynamism that could reduce the effectiveness of the static analysis. It may require a conservative overapproximation on what call sites can call what function bodies. If we wanted something like that, we could create a new kind of `Value` (`LambdaValue`?) to wrap dynamically-passed functions and pass them as arguments to other functions.

In that case, every `Operation` would have one non-optional `Operator` and the `Operator` would have an optional implementing `Module`, depending on whether the implementation is known. If we need dynamism like what you are describing, that could be indicated by an `Operator` that returns a `LambdaValue`, which may be passed to a function that calls that `LambdaValue`.

I believe having the implementing `Module` owned by the `Operator` could cause significant savings in memory, as static calls are usually the common case. There are generally many more call sites (`Operation`s) than function definitions (`Operator`s). Putting the `Module` on the `Operation` and pointing to it with a `unique_ptr` would require multiple "calls" to the same `Operator` to have duplicate copies of the implementation at different memory addresses, causing more space to be used and the cache to be hit less.

Reply:
I don't think this will necessarily reduce the effectiveness of static analysis. If the argument to a `CallFunction` is the result of `MakeFunction` operation, it is similar to a direct call? 

Also, I am not sure there will be a difference in memory usage.  There will be one `MakeFunction` operation which will hold on to `Module`. The `CallFunction` operation will refer to that instance of `MakeFunction` operation?  

In any case, there is so much to unpack here and is probably a topic for a broader discussion. Let us discuss this separately?

Reply:
Resolving this for now. We should have a longer conversation about this direction.